rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+
+*   should guarantee that the `BeamSource()`can return a valid
+    `PCollection[RecordBatch]` without a schema.
+    *   Other interfaces may raise an error when a schema was not provided.
+*   does not have to require a TFMD schema for all its interfaces to work.
+
+## (TensorFlow) Trainer integration
+
+For TFX to freely choose the storage format for training examples for a user, we
+**cannot** expose file-based or record-based interface to that user in the TF
+trainer, because:
+
+*   the user might not know how to open those files.
+*   there might not be an efficient representation of a “record” (this is true
+    for columnar storage formats like Apache Parquet) but only an efficient
+    representation of a batch of records.
+
+Thus we propose that to most users, the TF Trainer only exposes a handle to a
+`tf.data.Dataset` of parsed (composite) Tensors.
+
+Each `TFXIO` implementation will implement a `TFDataset()` interface to return
+such a `tf.data.Dataset`. This dataset contains logically a set of batched
+(composite) Tensors that are of the same type as the corresponding
+`TensorAdapter()` would return for a `RecordBatch`. See
+[this section](#recommended-way-of-implementing-a-tfxio) about how to minimize
+the code needs to be written for a new `TFXIO` implementation.
+
+The `TFDataset()` interface will accept common knobs that a user may need to
+tweak:
+
+*   Batch size
+*   Random shuffle
+
+## Code organization and OSS
+
+### `tfx_bsl` package
+
+TFXIO will be used by all TFX components as well as the TFX framework, making it
+be almost at the bottom of the dependency chain. Moreover, a lot of
+implementations details will be in C++, with python wrapping around, and we want
+to make sure our TFX components pip packages remain pure Python for easy
+maintenance. Therefore we propose a new python package
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) (TFX Shared Basic Libraries) to
+contain the implementations of `TFXIO` and other libraries shared across TFX
+components.
+
+![alt_text](20191017-tfx-standardized-inputs/oss_lib_org.png)
+
+## Recommended way of implementing a TFXIO
+
+To maximize code sharing, the following way of implementing a `TFXIO` is
+suggested:
+
+![alt_text](20191017-tfx-standardized-inputs/impl_tfxio.png)
+
+One would only need to implement the IO+Parsing-to-arrow in C++ once, and reuse
+it in the BeamSource() and a format-specific Dataset Op that produces a
+DT_VARIANT tensor that points to the parsed Arrow RecordBatch. Then we provide
+one C++ library that translates the Arrow RecordBatch to Tensors, which can also
+be reused in a TF op (as the downstream of the Dataset, or in a Python wrapper).
+
+# Alternatives Considered
+
+We’ve considered an alternative where
+[**StructuredTensor**](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md)
+is the unified in-memory representation, and **tf.Data** is the unified I/O
+abstraction.
```

Comment:
After reading this section, I'm actually less convinced that StructuredTensors aren't the answer here than when I started. Performance and convenience overheads could be minimized if necessary, and the question of requiring TF could be resolved by providing an inverse adapter-- rather than Arrow -> TF, convert TF -> Arrow on an as-needed basis. It seems like this route would simplify some of the build problems noted and reduce the total number of conversions required. @edloper and @alextp for StructuredTensors-- what would you argue for here?

Reply:
I was evaluating the options with the current state. Technically I believe we could bridge all the gaps mentioned above if we invest enough.

I'd like to argue that at the high level, even if the gap is bridged, there's still value to adopt Arrow rather than a TF construct:

1) TFX components do more data analysis than training a model. And that part does not have to depend on TF. Choosing a TF construct may subsequently mean choosing TF as the computation platform but I'm not sure TF is (or intends to be) the best data analysis platform. We might find us in a situation where our need (for data analysis) is not well aligned with where TF is going.

2) TFX components like TFDV and TFMA can still work without assuming TF as the ML library. And I believe TFX is moving towards the direction where supporting a different ML library is possible. Requiring users that don't use TF to learning about StructuredTensor seems to be weird. I'm not sure how much would an Arrow->StructuredTensor adapter would help, because they would still need to learn about both ends in order to do it correctly.

Reply:
I am strongly against having two slightly different ways of representing this type of structured data in the TF ecosystem. I also really wish these concerns were brought up during the review for structured tensor.

I also don't really understand the arguments here. A reasonable alternative to this document is "every tfx component uses tf ops to parse the data into a structuredtensor of eager tensors, and goes from there however it wants". That reuses only existing pieces, imposes no additional restrictions on the on-disk format of tfx components (since by definition it's anything that TF can read into a structured tensor, which will be anything TF can train a model over), and is fully configurable by the users. It also does not require a new arrow-like type in TF or even bidirectional conversions between structured tensors and arrow, though such conversions are relatively easy to write.

Reply:
I view it from a slightly different angle. This proposal is not introducing yet another way of representing structured data to the TF ecosystem (I assume the TF ecosystem is centered at TF, but TFX may not be (i.e. TF, the ML library in TFX, can potentially be replaced with other libraries). 

This proposal does not propose any new TF APIs, but only things to be built around TF, to work with TF, using existing APIs, and StructuredTensor is one of the APIs to be used.

/cc @katsiapis

Reply:
And again, unless we can use tf.data as a PSource in beam, tf.Dataset + StructuredTensor cannot fully replace this proposal.

Reply:
I added more background about TFX to the RFC (see section "Motivation" and "Requirements") to better describe the problem this proposal is trying to solve and what the workload that a TFX component / library typically bears. I also updated the "Alternatives considered" section with a head-to-head comparison. Hope this would clarify things.